lib/inventory/metadata/metadata.go

```diff
@@ -0,0 +1,334 @@
+type AgentMetadataFetchConfig struct {
+	Context context.Context
+	// getenv is the method called to retrieve an environment
+	// variable.
+	// It is configurable so that it can be mocked in tests.
+	getenv func(name string) string
+	// readFile is the method called to read a file.
+	// It is configurable so that it can be mocked in tests.
+	readFile func(name string) ([]byte, error)
+	// execCommand is the method called to execute a command.
+	// It is configurable so that it can be mocked in tests.
+	execCommand func(name string, args ...string) ([]byte, error)
+	// httpDo is the method called to perform an http request.
+	// It is configurable so that it can be mocked in tests.
+	httpDo func(req *http.Request) (*http.Response, error)
+	// kubeClient is a kubernetes client used to retrieve the
+	// server version.
+	// It is configurable so that it can be mocked in tests.
+	kubeClient kubernetes.Interface
+}
+
+// setDefaults sets the values of readFile and execCommand to the ones in the
+// standard library. Having these two methods configurable allows us to mock
+// them in tests.
+func (c *AgentMetadataFetchConfig) setDefaults() {
+	if c.Context == nil {
+		c.Context = context.Background()
+	}
+	if c.getenv == nil {
+		c.getenv = os.Getenv
+	}
+	if c.readFile == nil {
+		c.readFile = os.ReadFile
+	}
+	if c.execCommand == nil {
+		c.execCommand = func(name string, args ...string) ([]byte, error) {
+			return exec.Command(name, args...).Output()
+		}
+	}
+	if c.httpDo == nil {
+		c.httpDo = func(req *http.Request) (*http.Response, error) {
+			client, err := defaults.HTTPClient()
+			if err != nil {
+				return nil, err
+			}
+			client.Timeout = 5 * time.Second
+			return client.Do(req)
+		}
+	}
+	if c.kubeClient == nil {
+		c.kubeClient = getKubeClient()
+	}
+}
+
+// getKubeClient returns a kubernetes client in case the agent is running on
+// kubernetes. It returns nil otherwise.
+func getKubeClient() kubernetes.Interface {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		log.Debugf("Failed to get kubernetes cluster config: %s", err)
+		return nil
+	}
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		log.Debugf("Failed to create kubernetes client: %s", err)
+		return nil
+	}
+	return client
+}
+
+// FetchAgentMetadata fetches and calculates all agent metadata we are interested
+// in tracking.
+func FetchAgentMetadata(c *AgentMetadataFetchConfig) proto.UpstreamInventoryAgentMetadata {
+	c.setDefaults()
+	return proto.UpstreamInventoryAgentMetadata{
+		OS:                    c.fetchOS(),
+		OSVersionInfo:         c.fetchOSVersionInfo(),
+		HostArchitectureInfo:  c.fetchHostArchitectureInfo(),
+		GlibcVersionInfo:      c.fetchGlibcVersionInfo(),
+		InstallMethods:        c.fetchInstallMethods(),
+		ContainerRuntime:      c.fetchContainerRuntime(),
+		ContainerOrchestrator: c.fetchContainerOrchestrator(),
+		CloudEnvironment:      c.fetchCloudEnvironment(),
+	}
```

Comment:
Nit: I bet a lot of this info will be useful in contexts outside of just the metadata message/event.  Low-priority ask, but if you have some free time, it'd be nifty to have these individual fetch operations be a public API that we could leverage elsewhere in the codebase without doing a full fetch of all parameters.